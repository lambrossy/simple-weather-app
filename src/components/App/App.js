import { memo } from "react";
import styled from "styled-components";
import Grid from "../Grid/Grid";
import CityTile from "../CityTile/CityTile";
import useCities from "../../hooks/useCities";

const Container = styled.div({
  height: "100%",
  margin: "0 auto",
  maxWidth: 1200,
  minWidth: 600,
  display: "table",
  verticalAlign: "middle",
  color: "#48484a",
  fontFamily: '"Georama", sans-serif',
});

const Center = styled.div({
  display: "table-cell",
  verticalAlign: "middle",
});

const Message = styled.div({
  fontSize: 48,
  textAlign: "center",
});

const App = ({ dependencies }) => {
  const { cities, error, loading } = useCities(dependencies);

  return (
    <Container>
      <Center>
        {loading && <Message>Loading</Message>}
        {error && <Message>Whoops, something went wrong!</Message>}
        {cities && (
          <Grid>
            {cities.map((city) => (
              <CityTile key={city.id} {...city} />
            ))}
          </Grid>
        )}
      </Center>
    </Container>
  );
};

export default memo(App);
