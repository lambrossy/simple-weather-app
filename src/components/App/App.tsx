import { memo, FC } from "react";
import styled from "styled-components";
import Grid from "../Grid/Grid";
import CityTile from "../CityTile/CityTile";
import useCities from "../../hooks/useCities";
import CityWeather from "../../types/CityWeather";
import { AppDependencies } from "../../interfaces/AppDependencies";

const Container = styled.div({
  height: "100%",
  margin: "0 auto",
  maxWidth: 1200,
  minWidth: 600,
  display: "table",
  verticalAlign: "middle",
  color: "#48484a",
  fontFamily: '"Georama", sans-serif',
});

const Center = styled.div({
  display: "table-cell",
  verticalAlign: "middle",
});

const Message = styled.div({
  fontSize: 48,
  textAlign: "center",
});

interface Props {
  dependencies: AppDependencies;
}

const App: FC<Props> = ({ dependencies }) => {
  const { cities, error, loading } = useCities(dependencies);

  return (
    <Container>
      <Center>
        {loading && <Message>Loading</Message>}
        {error && <Message>Whoops, something went wrong!</Message>}
        {cities && (
          <Grid>
            {cities.map((city: CityWeather) => (
              <CityTile key={city.id} {...city} />
            ))}
          </Grid>
        )}
      </Center>
    </Container>
  );
};

export default memo(App);
