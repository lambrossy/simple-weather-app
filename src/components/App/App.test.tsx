import { render } from "@testing-library/react";
import * as SWR from "swr";
import App from "./App";

const dependencies = { fetch: () => Promise.resolve(true), env: {} };
const swr = {
  data: undefined,
  revalidate: () => Promise.resolve(false),
  mutate: () => Promise.resolve(false),
  isValidating: false,
};

const cities = [
  {
    id: 2158177,
    name: "Melbourne",
    description: "overcast clouds",
    temperature: { current: 13.06, max: 13.97, min: 11.14 },
    icon: {
      url: "http://openweathermap.org/img/wn/04n@2x.png",
      description: "Clouds",
    },
  },
  {
    id: 2147714,
    name: "Sydney",
    description: "clear sky",
    temperature: { current: 14.15, max: 16, min: 11.84 },
    icon: {
      url: "http://openweathermap.org/img/wn/01n@2x.png",
      description: "Clear",
    },
  },
  {
    id: 2174003,
    name: "Brisbane",
    description: "clear sky",
    temperature: { current: 15.25, max: 16.59, min: 12.94 },
    icon: {
      url: "http://openweathermap.org/img/wn/01n@2x.png",
      description: "Clear",
    },
  },
  {
    id: 2078025,
    name: "Adelaide",
    description: "clear sky",
    temperature: { current: 11.3, max: 13.25, min: 9.71 },
    icon: {
      url: "http://openweathermap.org/img/wn/01n@2x.png",
      description: "Clear",
    },
  },
  {
    id: 2172517,
    name: "Canberra",
    description: "overcast clouds",
    temperature: { current: 10.9, max: 11.82, min: 8.92 },
    icon: {
      url: "http://openweathermap.org/img/wn/04n@2x.png",
      description: "Clouds",
    },
  },
  {
    id: 2063523,
    name: "Perth",
    description: "clear sky",
    temperature: { current: 13.11, max: 15.44, min: 11.55 },
    icon: {
      url: "http://openweathermap.org/img/wn/01n@2x.png",
      description: "Clear",
    },
  },
  {
    id: 2073124,
    name: "Darwin",
    description: "clear sky",
    temperature: { current: 25.57, max: 25.99, min: 23.92 },
    icon: {
      url: "http://openweathermap.org/img/wn/01n@2x.png",
      description: "Clear",
    },
  },
  {
    id: 2163355,
    name: "Hobart",
    description: "few clouds",
    temperature: { current: 11.06, max: 11.81, min: 9.04 },
    icon: {
      url: "http://openweathermap.org/img/wn/02n@2x.png",
      description: "Clouds",
    },
  },
];

it("renders loading state correctly", () => {
  jest.spyOn(SWR, "default").mockReturnValue(swr);
  const screen = render(<App dependencies={dependencies} />);
  expect(screen.asFragment()).toMatchSnapshot();
});

it("renders error state correctly", () => {
  jest.spyOn(SWR, "default").mockReturnValue({ ...swr, error: true });
  const screen = render(<App dependencies={dependencies} />);
  expect(screen.asFragment()).toMatchSnapshot();
});

it("renders success state correctly", () => {
  jest.spyOn(SWR, "default").mockReturnValue({ ...swr, data: cities });
  const screen = render(<App dependencies={dependencies} />);
  expect(screen.asFragment()).toMatchSnapshot();
});
