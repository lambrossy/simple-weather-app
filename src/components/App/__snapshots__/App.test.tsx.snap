// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error state correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-jrsJWt ZZLqs"
  >
    <div
      class="sc-kEqXSa fBXjEc"
    >
      <div
        class="sc-iqAclL gIyUmm"
      >
        Whoops, something went wrong!
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders loading state correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-jrsJWt ZZLqs"
  >
    <div
      class="sc-kEqXSa fBXjEc"
    >
      <div
        class="sc-iqAclL gIyUmm"
      >
        Loading
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders success state correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-jrsJWt ZZLqs"
  >
    <div
      class="sc-kEqXSa fBXjEc"
    >
      <div
        class="sc-bdnxRM fwJTUD"
      >
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Melbourne
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  13°
                </div>
                <img
                  alt="Clouds"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/04n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Overcast clouds
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 11°
                </div>
                <div>
                  High: 14°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Sydney
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  14°
                </div>
                <img
                  alt="Clear"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/01n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Clear sky
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 12°
                </div>
                <div>
                  High: 16°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Brisbane
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  15°
                </div>
                <img
                  alt="Clear"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/01n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Clear sky
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 13°
                </div>
                <div>
                  High: 17°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Adelaide
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  11°
                </div>
                <img
                  alt="Clear"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/01n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Clear sky
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 10°
                </div>
                <div>
                  High: 13°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Canberra
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  11°
                </div>
                <img
                  alt="Clouds"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/04n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Overcast clouds
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 9°
                </div>
                <div>
                  High: 12°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Perth
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  13°
                </div>
                <img
                  alt="Clear"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/01n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Clear sky
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 12°
                </div>
                <div>
                  High: 15°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Darwin
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  26°
                </div>
                <img
                  alt="Clear"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/01n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Clear sky
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 24°
                </div>
                <div>
                  High: 26°
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gtsrHT jnDyyg"
        >
          <div
            class="sc-dlnjwi eSZeum"
          >
            <div
              class="sc-hKFxyN bAHqqK"
            >
              <h2
                class="sc-eCApnc gitqnN"
              >
                Hobart
              </h2>
              <div
                class="sc-gKAaRy gPYLQA"
              >
                <div
                  class="sc-jSFjdj cAlUKH"
                >
                  11°
                </div>
                <img
                  alt="Clouds"
                  class="sc-iCoGMd hcYqGP"
                  src="http://openweathermap.org/img/wn/02n@2x.png"
                />
              </div>
              <h2
                class="sc-fujyAs bNiUUU"
              >
                Few clouds
              </h2>
              <div
                class="sc-pNWdM jbXcIf"
              >
                <div>
                  Low: 9°
                </div>
                <div>
                  High: 12°
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
